Introduce user profiles

azasypkin opened the pull request:
## Summary

This PR merges `feature/user-profile` branch into `main` branch.

~~__Blocked by: https://github.com/elastic/kibana/issues/132645__~~

------------------

Release note: Users can have their personal avatars now.

elasticmachine commented:
Pinging @elastic/kibana-security (Team:Security)

azasypkin commented:
We might want to merge/review the following PR before merging `feature/user-profiles` to `main`: https://github.com/elastic/kibana/pull/132622

jbudz reviewed (approved):
limits.yml

MichaelMarcialis commented:
@azasypkin, @thomheymann: Left a question for ya'll in the [previous PR](https://github.com/elastic/kibana/pull/127624#issuecomment-1136297977) before I review this.

azasypkin commented:
@elastic/kibana-core the Core-owned changes in this PR are exactly the same you've already reviewed in https://github.com/elastic/kibana/pull/132622. The only difference is that https://github.com/elastic/kibana/pull/132622 was targeting a feature branch and with this PR we're going to merge to `main`.

azasypkin commented on x-pack/plugins/security/public/components/form_label.tsx:
```diff
@@ -0,0 +1,58 @@
+  useEffect(() => report(isEqual), [isEqual]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  return (
+    <EuiFlexGroup responsive={false} gutterSize="xs">
+      <EuiFlexItem grow={false}>{props.children}</EuiFlexItem>
+      {!isEqual ? (
+        <EuiFlexItem grow={false}>
+          <EuiIcon type="dot" color="success" />
+        </EuiFlexItem>
+      ) : undefined}
+    </EuiFlexGroup>
```
I must be doing something wrong, but I still see the jump even with `verticalAlign: 'text-top'`: https://user-images.githubusercontent.com/1713708/171826635-13ab94ca-90db-475f-98a2-158689abc930.mp4 (on the left is what we have with flex layout, and on the right - the markup below).

```
    <>
      {props.children}{' '}
      {!isEqual ? (
        <EuiIcon type="dot" color="success" style={{ verticalAlign: 'text-top' }} />
      ) : undefined}
    </>
```

MichaelMarcialis replied:
Maybe I'm missing it, but it looks to be corrected in your video when the suggested vertical alignment style is applied to the icon (and then the issue returns when it's disabled). If I'm mistaken however, and the vertical alignment of the icon isn't working, then feel free to ignore for now. It's a minor issue.

azasypkin commented:
> It looks like the padding I previously requested is being applied individually to the EuiPageHeader and EuiPageContentBody components, rather being applied to the parent page template. This is noticeable by the border not being aligned with the text. What component is being used for the page template currently? It doesn't look like KibanaPageTemplate (which functions as a sort of extension of EUI's EuiPageTemplate) from kbn-shared-ux-components is being used as we do in other areas of Kibana. Should we be using that component here? The padding should already enabled by default on that component.

You're absolutely right, we didn't use `KibanaPageTemplate`. I changed that, but it doesn't seem to help when we set `KibanaPageTemplate.pageHeader.bottomBorder` to `true` since both default paddings and border are applied to the same `.euiPageHeader` container. To fix this, I had to restore `{ paddingLeft: 0, paddingRight: 0 }` for `EuiPageHeader` and `EuiPageContentBody` I added previously in https://github.com/elastic/kibana/pull/127624.

> Can we change the subdued color on the "None provided" text to use $euiColorDisabledText color (i.e. #abb4c4) instead to better match the designs?

Changed to `$euiColorDisabled` (I assume you meant this one and not `$euiColorDisabledText`) in https://github.com/elastic/kibana/pull/132522/commits/92988cd9aa09d4074196e4ae0ab5757cc3f13223

azasypkin commented:
> "page load bundle size" for "security" is 25.8KB over the configured limit of 97.7KB 

Dependency on `kbn-shared-ux-components` leads to a 25%+ increase in our page bundle size :cry: And since it doesn't give us anything yet, I'll switch back to the raw `EuiPageTemplate`.

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @azasypkin! Leaving you two final comments for the re-review:

#### Disabled text color

> Changed to `$euiColorDisabled` (I assume you meant this one and not `$euiColorDisabledText`) in [92988cd](https://github.com/elastic/kibana/commit/92988cd9aa09d4074196e4ae0ab5757cc3f13223)

They're actually the same hex values, but I did indeed intend to use `$euiColorDisabledText` (or `euiTheme.colors.disabledText`), just in case the EUI team changes the value specifically for the text version in the future. Can we change that?

#### Page template padding

With the swapping that happened between `KibanaPageTemplate` and `EuiPageTemplate`, I think the padding got messed up again in the process. It looks like there is no padding being applied now, apparent when viewing at small viewport sizes.

![image](https://user-images.githubusercontent.com/3884767/171940026-d2b9af05-23f8-4539-b774-499baa9629f3.png)

I'm guessing this is happening because there is a `paddingSize="none"` prop being applied to the `EuiPageTemplate`. If so, can you remove that prop (which should theoretically apply the default `paddingSize="l"`)?

Otherwise, this looks good to me. I'm going to go ahead and approve it now so I don't hold you up any further, under the assumption that you can address the above comments. Thanks so much!

azasypkin commented:
> They're actually the same hex values, but I did indeed intend to use $euiColorDisabledText (or euiTheme.colors.disabledText), just in case the EUI team changes the value specifically for the text version in the future. Can we change that?

Hmm, when I use `euiTheme.colors.disabledText` I get `rgb(162, 171, 186)` (`#a2abba`), but with `euiTheme.colors.disabled` I get `rgb(171, 180, 196)` (`#abb4c4`, the one we have in design mockups). Looking at this code [this code](https://github.com/elastic/eui/blob/e0f9876bfbbf33414d494acb33478ec702c4cc17/src/themes/amsterdam/global_styling/variables/_colors.ts#L61-L62) I assumed it was intentional, or not @MichaelMarcialis ?

azasypkin commented on x-pack/plugins/security/public/account_management/user_profile/utils.ts:
```diff
@@ -0,0 +1,76 @@
+      }
+
+      try {
+        const canvas = document.createElement('canvas');
+        const context = canvas.getContext('2d');
+        if (context) {
+          if (image.width >= image.height) {
+            canvas.width = maxSize;
+            canvas.height = Math.floor((image.height * maxSize) / image.width);
+          } else {
+            canvas.height = maxSize;
+            canvas.width = Math.floor((image.width * maxSize) / image.height);
+          }
+          context.drawImage(image, 0, 0, canvas.width, canvas.height);
+          const resizedDataUrl = canvas.toDataURL();
```
Funnily enough, centering "tall" images works much worse for the image set I have since the head in tall photos is usually close to (0,0) and centering clips the head sometimes like in the picture below:

![image](https://user-images.githubusercontent.com/1713708/172579372-1adb51ea-d6d1-47c4-ad77-0543ac7b05f5.png)

I'll keep as is for now, and we can revisit it with @thomheymann once he's back

azasypkin commented on x-pack/plugins/security/server/authentication/authentication_result.ts:
```diff
@@ -49,8 +50,21 @@ interface AuthenticationOptions {
+  userProfileGrant?: UserProfileGrant;
 }
 
+export type SucceededAuthenticationResultOptions = Pick<
+  AuthenticationOptions,
+  'authHeaders' | 'userProfileGrant' | 'authResponseHeaders' | 'state'
+>;
+
+export type RedirectedAuthenticationResultOptions = Pick<
+  AuthenticationOptions,
+  'user' | 'userProfileGrant' | 'authResponseHeaders' | 'state'
+>;
+
+export type FailedAuthenticationResultOptions = Pick<AuthenticationOptions, 'authResponseHeaders'>;
```
I believe these types aren't not included in our public API, but I agree - `Pick` is handy, but harmful - I'll get rid of it.

azasypkin commented on x-pack/plugins/security/server/routes/user_profile/get.ts:
```diff
@@ -0,0 +1,52 @@
+      }
+
+      if (!session.userProfileId) {
+        logger.warn(`User profile missing from current session. (sid: ${session.sid.slice(-10)})`);
```
Handled in https://github.com/elastic/kibana/pull/132522/commits/6442c8cc5de7a1b94e169c7a5c3c9896376c9ed8

azasypkin commented on x-pack/plugins/security/public/account_management/user_profile/user_avatar.tsx:
```diff
@@ -0,0 +1,50 @@
+import type { FunctionComponent, HTMLAttributes } from 'react';
+import React from 'react';
+
+import type { UserAvatar as IUserAvatar, UserInfo } from '../../../common';
```
Handled in https://github.com/elastic/kibana/pull/132522/commits/6442c8cc5de7a1b94e169c7a5c3c9896376c9ed8

azasypkin commented on x-pack/plugins/security/public/components/api_clients_provider.ts:
```diff
@@ -0,0 +1,23 @@
+  users,
+}));
+
+export { ApiClientsProvider, useApiClients };
```
Handled in https://github.com/elastic/kibana/pull/132522/commits/6442c8cc5de7a1b94e169c7a5c3c9896376c9ed8

azasypkin commented on x-pack/plugins/security/public/account_management/user_profile/user_profile.tsx:
```diff
@@ -0,0 +1,701 @@
+                </EuiDescribedFormGroup>
+              ) : null}
+
+              {formik.values.data ? (
```
Handled in https://github.com/elastic/kibana/pull/132522/commits/6442c8cc5de7a1b94e169c7a5c3c9896376c9ed8

azasypkin commented:
> I'm not sure if you intended to address https://github.com/elastic/kibana/issues/132645 before merging this to main or not, but I just wanted to remind you of it.
If you want to follow up afterwards, that's fine.

Thanks for the reminder! Yeah, we agreed to handle most of the feedback recorded in https://github.com/elastic/kibana/issues/132645 separately.

MichaelMarcialis commented:
> Hmm, when I use `euiTheme.colors.disabledText` I get `rgb(162, 171, 186)` (`#a2abba`), but with `euiTheme.colors.disabled` I get `rgb(171, 180, 196)` (`#abb4c4`, the one we have in design mockups). Looking at this code [this code](https://github.com/elastic/eui/blob/e0f9876bfbbf33414d494acb33478ec702c4cc17/src/themes/amsterdam/global_styling/variables/_colors.ts#L61-L62) I assumed it was intentional, or not @MichaelMarcialis ?

Ah, you're right. I was looking at the old Sass variables. Some of the newer Emotion colors are calculated slightly differently than they were previously and we still need to update the design system in Figma to get them back in sync. Thanks for catching this. In any case, let's go ahead and proceed with `euiTheme.colors.disabledText` here.

azasypkin commented:
> In any case, let's go ahead and proceed with euiTheme.colors.disabledText here.

Roger that, switched to `disabledText` in https://github.com/elastic/kibana/pull/132522/commits/dda0a109681c44e07453d20be71cff1d5c773a3c, thanks.
